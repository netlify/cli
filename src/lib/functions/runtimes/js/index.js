const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda')
const detectZisiBuilder = require('./builders/zisi')

const SECONDS_TO_MILLISECONDS = 1e3

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
