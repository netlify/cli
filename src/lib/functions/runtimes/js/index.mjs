import { dirname } from 'path'

import lambdaLocal from '@skn0tt/lambda-local'
import winston from 'winston'

import detectNetlifyLambdaBuilder from './builders/netlify-lambda.mjs'
import detectZisiBuilder, { parseForSchedule } from './builders/zisi.mjs'

export const name = 'js'

const SECONDS_TO_MILLISECONDS = 1000

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

export const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

export const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

export const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}
