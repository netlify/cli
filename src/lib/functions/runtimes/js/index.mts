// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'dirname'.
const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda.cjs')
const detectZisiBuilder = require('./builders/zisi.cjs')

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'SECONDS_TO... Remove this comment to see the full error message
const SECONDS_TO_MILLISECONDS = 1e3

// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
let netlifyLambdaDetectorCache: $TSFixMe

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'getBuildFu... Remove this comment to see the full error message
const getBuildFunction = async ({
  config,
  directory,
  errorExit,
  func,
  projectRoot
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'invokeFunc... Remove this comment to see the full error message
const invokeFunction = async ({
  context,
  event,
  func,
  timeout
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
