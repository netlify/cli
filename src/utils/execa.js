const { env } = require('process')

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).
// eslint-disable-next-line import/no-dynamic-require
const execa = require(env.NETLIFY_CLI_EXECA_PATH || 'execa')

module.exports = execa
