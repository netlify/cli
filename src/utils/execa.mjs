import { env } from 'process'
// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).
// eslint-disable-next-line import/no-mutable-exports
let execa

if (env.NETLIFY_CLI_EXECA_PATH) {
  // eslint-disable-next-line import/no-dynamic-require
  const execaMock = await import(env.NETLIFY_CLI_EXECA_PATH)
  execa = execaMock.default
} else {
  const execaLib = await import('execa')
  execa = execaLib.default
}

export default execa
