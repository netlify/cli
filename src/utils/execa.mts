const { env } = require('process')

const execaLib = require('execa')

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).

// eslint-disable-next-line import/no-dynamic-require
const execa = env.NETLIFY_CLI_EXECA_PATH ? require(env.NETLIFY_CLI_EXECA_PATH) : execaLib

module.exports = execa
