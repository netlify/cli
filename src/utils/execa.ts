import { env } from 'process'

import execaLib from 'execa'

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).

/**
 * @type {import('execa')}
 */
// eslint-disable-next-line import/no-mutable-exports
let execa

if (env.NETLIFY_CLI_EXECA_PATH) {
  const execaMock = await import(env.NETLIFY_CLI_EXECA_PATH)
  execa = execaMock.default
} else {
  execa = execaLib
}

export default execa
