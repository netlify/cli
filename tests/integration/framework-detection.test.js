import execa from 'execa'
import fetch from 'node-fetch'
import { describe, test } from 'vitest'

import { cliPath } from './utils/cli-path.js'
import { getExecaOptions, withDevServer } from './utils/dev-server.ts'
import { DOWN, answerWithValue, handleQuestions } from './utils/handle-questions.js'
import { withSiteBuilder } from './utils/site-builder.ts'
import { normalize } from './utils/snapshots.js'

const content = 'Hello World!'

describe.concurrent('frameworks/framework-detection', () => {
  test('should default to process.cwd() and static server', async (t) => {
    await withSiteBuilder('site-with-index-file', async (builder) => {
      await builder
        .withContentFile({
          path: 'index.html',
          content,
        })
        .buildAsync()

      await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
        const response = await fetch(url).then((res) => res.text())
        t.expect(response).toEqual(content)

        t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
      })
    })
  })

  test('should use static server when --dir flag is passed', async (t) => {
    await withSiteBuilder('site-with-index-file', async (builder) => {
      await builder
        .withContentFile({
          path: 'public/index.html',
          content,
        })
        .buildAsync()

      await withDevServer({ cwd: builder.directory, args: ['--dir', 'public'] }, async ({ output, url }) => {
        const response = await fetch(url).then((res) => res.text())
        t.expect(response).toEqual(content)

        t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
      })
    })
  })

  test('should use static server when framework is set to #static', async (t) => {
    await withSiteBuilder('site-with-index-file', async (builder) => {
      await builder
        .withContentFile({
          path: 'index.html',
          content,
        })
        .withNetlifyToml({ config: { dev: { framework: '#static' } } })
        .buildAsync()

      await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
        const response = await fetch(url).then((res) => res.text())
        t.expect(response).toEqual(content)

        t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
      })
    })
  })

  // This test has a race condition that occasionally causes it to fail when run concurrently.
  // Running it in isolation (or removing the '.concurrent' on the describe block above)
  // fixes it. See CT-1094 for more details
  test('should log the command if using static server and `command` is configured', async (t) => {
    await withSiteBuilder('site-with-index-file', async (builder) => {
      await builder
        .withContentFile({
          path: 'public/index.html',
          content,
        })
        .buildAsync()

      await withDevServer(
        { cwd: builder.directory, args: ['--dir', 'public', '--command', 'npm run start'] },
        async ({ output, url }) => {
          const response = await fetch(url).then((res) => res.text())
          t.expect(response).toEqual(content)

          t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
        },
      )
    })
  })

  test('should warn if using static server and `targetPort` is configured', async (t) => {
    await withSiteBuilder('site-with-index-file', async (builder) => {
      await builder
        .withContentFile({
          path: 'public/index.html',
          content,
        })
        .buildAsync()

      await withDevServer(
        { cwd: builder.directory, args: ['--dir', 'public', '--target-port', '3000'] },
        async ({ output, url }) => {
          const response = await fetch(url).then((res) => res.text())
          t.expect(response).toEqual(content)

          t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
        },
      )
    })
  })

  test('should run `command` when both `command` and `targetPort` are configured', async (t) => {
    await withSiteBuilder('empty-site', async (builder) => {
      await builder.withNetlifyToml({ config: { build: { publish: 'public' } } }).buildAsync()

      // a failure is expected since we use `echo hello` instead of starting a server
      const error = await withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--target-port', '3000'] },
        () => {},
        true,
      ).catch((error_) => error_)

      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should force a specific framework when configured', async (t) => {
    await withSiteBuilder('site-with-mocked-cra', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: 'create-react-app' } } }).buildAsync()

      // a failure is expected since this is not a true create-react-app project
      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should throw when forcing a non supported framework', async (t) => {
    await withSiteBuilder('site-with-unknown-framework', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: 'to-infinity-and-beyond-js' } } }).buildAsync()

      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should detect a known framework', async (t) => {
    await withSiteBuilder('site-with-cra', async (builder) => {
      await builder
        .withPackageJson({
          packageJson: { dependencies: { 'react-scripts': '1.0.0' }, scripts: { start: 'react-scripts start' } },
        })
        .buildAsync()

      // a failure is expected since this is not a true create-react-app project
      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should throw if framework=#custom but command is missing', async (t) => {
    await withSiteBuilder('site-with-framework-and-no-command', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: '#custom' } } }).buildAsync()

      const error = await withDevServer(
        { cwd: builder.directory, args: ['--target-port', '3000'] },
        () => {},
        true,
      ).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should throw if framework=#custom but targetPort is missing', async (t) => {
    await withSiteBuilder('site-with-framework-and-no-command', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: '#custom' } } }).buildAsync()

      const error = await withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello'] },
        () => {},
        true,
      ).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should start custom command if framework=#custom, command and targetPort are configured', async (t) => {
    await withSiteBuilder('site-with-custom-framework', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: '#custom', publish: 'public' } } }).buildAsync()

      const error = await withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--target-port', '3000'] },
        () => {},
        true,
      ).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test(`should print specific error when command doesn't exist`, async (t) => {
    await withSiteBuilder('site-with-custom-framework', async (builder) => {
      await builder.buildAsync()

      const error = await withDevServer(
        {
          cwd: builder.directory,
          args: [
            '--command',
            'oops-i-did-it-again forgot-to-use-a-valid-command',
            '--target-port',
            '3000',
            '--framework',
            '#custom',
          ],
        },
        () => {},
        true,
      ).catch((error_) => error_)

      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should prompt when multiple frameworks are detected', async (t) => {
    await withSiteBuilder('site-with-multiple-frameworks', async (builder) => {
      await builder
        .withPackageJson({
          packageJson: {
            dependencies: { 'react-scripts': '1.0.0', gatsby: '^3.0.0' },
            scripts: { start: 'react-scripts start', develop: 'gatsby develop' },
          },
        })
        .withContentFile({ path: 'gatsby-config.js', content: '' })
        .buildAsync()

      // a failure is expected since this is not a true framework project
      const asyncErrorBlock = async () => {
        const childProcess = execa(cliPath, ['dev', '--offline'], getExecaOptions({ cwd: builder.directory }))

        handleQuestions(childProcess, [
          {
            question: 'Multiple possible dev commands found',
            answer: answerWithValue(DOWN),
          },
        ])

        await childProcess
      }
      const error = await asyncErrorBlock().catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should not run framework detection if command and targetPort are configured', async (t) => {
    await withSiteBuilder('site-with-hugo-config', async (builder) => {
      await builder.withContentFile({ path: 'config.toml', content: '' }).buildAsync()

      // a failure is expected since the command exits early
      const error = await withDevServer(
        { cwd: builder.directory, args: ['--command', 'echo hello', '--target-port', '3000'] },
        () => {},
        true,
      ).catch((error_) => error_)

      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should filter frameworks with no dev command', async (t) => {
    await withSiteBuilder('site-with-gulp', async (builder) => {
      await builder
        .withContentFile({
          path: 'index.html',
          content,
        })
        .withPackageJson({
          packageJson: { dependencies: { gulp: '1.0.0' } },
        })
        .buildAsync()

      await withDevServer({ cwd: builder.directory }, async ({ output, url }) => {
        const response = await fetch(url).then((res) => res.text())
        t.expect(response).toEqual(content)

        t.expect(normalize(output, { duration: true, filePath: true })).toMatchSnapshot()
      })
    })
  })

  test('should pass framework-info env to framework sub process', async (t) => {
    await withSiteBuilder('site-with-gatsby', async (builder) => {
      await builder
        .withPackageJson({
          packageJson: {
            dependencies: { nuxt3: '^2.0.0' },
            scripts: { dev: 'node -p process.env.NODE_VERSION' },
          },
        })
        .buildAsync()

      // a failure is expected since this is not a true Gatsby project
      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(normalize(error.stdout, { duration: true, filePath: true })).toMatchSnapshot()
    })
  })

  test('should start static service for frameworks without port, forced framework', async (t) => {
    await withSiteBuilder('site-with-remix', async (builder) => {
      await builder.withNetlifyToml({ config: { dev: { framework: 'remix' } } }).buildAsync()

      // a failure is expected since this is not a true remix project
      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(error.stdout.includes(`Failed running command: remix watch. Please verify 'remix' exists`)).toBe(true)
    })
  })

  test('should start static service for frameworks without port, detected framework', async (t) => {
    await withSiteBuilder('site-with-remix', async (builder) => {
      await builder
        .withPackageJson({
          packageJson: {
            dependencies: { remix: '^1.0.0', '@remix-run/netlify': '^1.0.0' },
            scripts: {},
          },
        })
        .withContentFile({ path: 'remix.config.js', content: '' })
        .buildAsync()

      // a failure is expected since this is not a true remix project
      const error = await withDevServer({ cwd: builder.directory }, () => {}, true).catch((error_) => error_)
      t.expect(error.stdout.includes(`Failed running command: remix watch. Please verify 'remix' exists`)).toBe(true)
    })
  })

  test('should run and serve a production build when using the `serve` command', async (t) => {
    await withSiteBuilder('site-with-framework', async (builder) => {
      await builder
        .withNetlifyToml({
          config: {
            build: { publish: 'public' },
            context: {
              dev: { environment: { CONTEXT_CHECK: 'DEV' } },
              production: { environment: { CONTEXT_CHECK: 'PRODUCTION' } },
            },
            functions: { directory: 'functions' },
            plugins: [{ package: './plugins/frameworker' }],
          },
        })
        .withBuildPlugin({
          name: 'frameworker',
          plugin: {
            onPreBuild: async ({ netlifyConfig }) => {
              // eslint-disable-next-line n/global-require, no-undef
              const { mkdir, writeFile } = require('fs/promises')

              const generatedFunctionsDir = 'new_functions'
              netlifyConfig.functions.directory = generatedFunctionsDir

              netlifyConfig.redirects.push({
                from: '/hello',
                to: '/.netlify/functions/hello',
              })

              await mkdir(generatedFunctionsDir)
              await writeFile(
                `${generatedFunctionsDir}/hello.js`,
                `const { CONTEXT_CHECK, NETLIFY_DEV } = process.env; exports.handler = async () => ({ statusCode: 200, body: JSON.stringify({ CONTEXT_CHECK, NETLIFY_DEV }) })`,
              )
            },
          },
        })
        .buildAsync()

      await withDevServer(
        { cwd: builder.directory, context: null, debug: true, serve: true },
        async ({ output, url }) => {
          const response = await fetch(`${url}/hello`).then((res) => res.json())
          t.expect(response).toStrictEqual({ CONTEXT_CHECK: 'PRODUCTION' })

          const normalizedText = normalize(output, { duration: true, filePath: true })
          t.expect(
            normalizedText.includes(
              `Changes will not be hot-reloaded, so if you need to rebuild your site you must exit and run 'netlify serve' again`,
            ),
          ).toEqual(true)
        },
      )
    })
  })
})
