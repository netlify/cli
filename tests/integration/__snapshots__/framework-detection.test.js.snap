// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`frameworks/framework-detection > should default to process.cwd() and static server 1`] = `
"◈ Netlify Dev ◈
◈ No app server detected. Using simple static server
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://docs.netlify.com/cli/local-development/#project-detection
◈ Running static server from \\"should-default-to-process-cwd-and-static-server\\"
◈ Setting up local development server

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘"
`;

exports[`frameworks/framework-detection > should detect a known framework 1`] = `
"◈ Netlify Dev ◈
◈ Setting up local development server
◈ Starting Netlify Dev with Create React App

> start
> react-scripts start

◈ Command failed with exit code *: npm run start. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should filter frameworks with no dev command 1`] = `
"◈ Netlify Dev ◈
◈ No app server detected. Using simple static server
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://docs.netlify.com/cli/local-development/#project-detection
◈ Running static server from \\"should-filter-frameworks-with-no-dev-command\\"
◈ Setting up local development server

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘"
`;

exports[`frameworks/framework-detection > should force a specific framework when configured 1`] = `
"◈ Netlify Dev ◈
◈ Setting up local development server
◈ Starting Netlify Dev with Create React App
◈ Failed running command: react-scripts start. Please verify 'react-scripts' exists"
`;

exports[`frameworks/framework-detection > should not run framework detection if command and targetPort are configured 1`] = `
"◈ Netlify Dev ◈
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://docs.netlify.com/cli/local-development/#project-detection
◈ Setting up local development server
◈ Starting Netlify Dev with custom config
hello
◈ \\"echo hello\\" exited with code *. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should pass framework-info env to framework sub process 1`] = `
"◈ Netlify Dev ◈
◈ Setting up local development server
◈ Starting Netlify Dev with Nuxt 3

> dev
> node -p process.env.NODE_VERSION

18
◈ \\"npm run dev\\" exited with code *. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should print specific error when command doesn't exist 1`] = `
"◈ Netlify Dev ◈
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://docs.netlify.com/cli/local-development/#project-detection
◈ Setting up local development server
◈ Starting Netlify Dev with #custom
◈ Failed running command: oops-i-did-it-again forgot-to-use-a-valid-command. Please verify 'oops-i-did-it-again' exists"
`;

exports[`frameworks/framework-detection > should prompt when multiple frameworks are detected 1`] = `
"◈ Netlify Dev ◈
? Multiple possible dev commands found (Use arrow keys or type to search)
> [Gatsby] 'npm run develop' 
  [Create React App] 'npm run start' [2A[37D[73C[2B[2K[1A[2K[1A[2K[G? Multiple possible dev commands found 
  [Gatsby] 'npm run develop' 
> [Create React App] 'npm run start' [2A[37D[39C[2B[2K[1A[2K[1A[2K[G? Multiple possible dev commands found Create React App-npm run start[69D[69C

Update your netlify.toml to avoid this selection prompt next time:

[build]
command = \\"react-scripts build\\"
publish = \\"build\\"

[dev]
command = \\"npm run start\\"

◈ Setting up local development server
◈ Starting Netlify Dev with Create React App

> start
> react-scripts start

◈ Command failed with exit code *: npm run start. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should run \`command\` when both \`command\` and \`targetPort\` are configured 1`] = `
"◈ Netlify Dev ◈
◈ Setting up local development server
◈ Starting Netlify Dev with custom config
hello
◈ \\"echo hello\\" exited with code *. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should start custom command if framework=#custom, command and targetPort are configured 1`] = `
"◈ Netlify Dev ◈
◈ Setting up local development server
◈ Starting Netlify Dev with #custom
hello
◈ \\"echo hello\\" exited with code *. Shutting down Netlify Dev server"
`;

exports[`frameworks/framework-detection > should throw if framework=#custom but command is missing 1`] = `
"◈ Netlify Dev ◈
◈ 'command' and 'targetPort' properties are required when 'framework' is set to '#custom'"
`;

exports[`frameworks/framework-detection > should throw if framework=#custom but targetPort is missing 1`] = `
"◈ Netlify Dev ◈
◈ 'command' and 'targetPort' properties are required when 'framework' is set to '#custom'"
`;

exports[`frameworks/framework-detection > should throw when forcing a non supported framework 1`] = `
"◈ Netlify Dev ◈
◈ Invalid framework \\"to-infinity-and-beyond-js\\". It should be one of: *"
`;

exports[`frameworks/framework-detection > should use static server when --dir flag is passed 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '--dir' flag was specified
◈ Running static server from \\"should-use-static-server-when-dir-flag-is-passed/public\\"
◈ Setting up local development server

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘"
`;

exports[`frameworks/framework-detection > should use static server when framework is set to #static 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '[dev.framework]' was set to '#static'
◈ Unable to determine public folder to serve files from. Using current working directory
◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.
◈ See docs at: https://docs.netlify.com/cli/local-development/#project-detection
◈ Running static server from \\"should-use-static-server-when-framework-is-set-to-static\\"
◈ Setting up local development server

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘"
`;

exports[`frameworks/framework-detection > should warn if using static server and \`targetPort\` is configured 1`] = `
"◈ Netlify Dev ◈
◈ Using simple static server because '--dir' flag was specified
◈ Ignoring 'targetPort' setting since using a simple static server.
◈ Use --staticServerPort or [dev.staticServerPort] to configure the static server port
◈ Running static server from \\"should-warn-if-using-static-server-and-target-port-is-configured/public\\"
◈ Setting up local development server

◈ Static server listening to 88888

   ┌──────────────────────────────────────────────────┐
   │                                                  │
   │   ◈ Server now ready on http://localhost:88888   │
   │                                                  │
   └──────────────────────────────────────────────────┘"
`;
